src/data/int/modeq.lean

```diff
@@ -146,13 +155,40 @@ by rw [modeq_iff_dvd] at *; exact (dvd_mul_left n m).trans h
 theorem of_mul_right (m : ℤ) : a ≡ b [ZMOD n * m] → a ≡ b [ZMOD n] :=
 mul_comm m n ▸ of_mul_left _
 
+/-- To cancel a common factor `c` from a `modeq` we must divide the modulus `m` by `gcd m c`. -/
+lemma cancel_left_div_gcd (hm : 0 < m) (h : c * a ≡ c * b [ZMOD m]) : a ≡ b [ZMOD m / gcd m c] :=
+begin
+  let d := gcd m c,
+  have hmd := gcd_dvd_left m c,
+  have hcd := gcd_dvd_right m c,
+  rw modeq_iff_dvd at ⊢ h,
+  refine int.dvd_of_dvd_mul_right_of_gcd_one _ _,
+  show m / d ∣ c / d * (b - a),
+  { rw [mul_comm, ←int.mul_div_assoc (b - a) hcd, mul_comm],
+    apply int.div_dvd_div hmd,
+    rwa mul_sub },
+  { rw [gcd_div hmd hcd, nat_abs_of_nat, nat.div_self (gcd_pos_of_ne_zero_left c hm.ne')] }
+end
+
+lemma cancel_right_div_gcd (hm : 0 < m) (h : a * c ≡ b * c [ZMOD m]) : a ≡ b [ZMOD m / gcd m c] :=
+by { apply cancel_left_div_gcd hm, simpa [mul_comm] using h }
```

Comment:
nit: to me, the `right` lemma is the more natural one as the `c` is on the same side of the `m`, `a`, and `b`. Of course we should still have both, but I'd prefer to see them proved in terms of each other in the other order.

Reply:
I guess by similar reasoning you could possibly argue that the `gcd m c` should be `gcd c m` in the left version.

Reply:
I would rather keep `gcd m c` in both because `m` is supposed to be an external parameter.

Reply:
External in what sense? I'm envisaging a situation where you apply these two lemmas to `d * a * c ≡ d * b * c [ZMOD m]`, and getting a a result that loosely reflects the positioning of `d` and `c`.

I guess with your version you get `(m / gcd m d) / gcd (m / gcd m d) c` and mine you get `(m / gcd d m / gcd (m / gcd d m) c`.

Either way,

bors d+

Reply:
Oh I see. I'll say it doesn't matter much. We can revisit later if we feel the need to.